CesiumGS/cesium
# DX: Fix `defined` and `Check` types for CesiumJS devs

# Description

- Add `defined.d.ts` which will now get picked up and provide the correct type predicates and assertions so JS Type checking will understand things are defined after this call
    - Originally added in https://github.com/CesiumGS/cesium/pull/11455 which is fantastic for _external_ contributors but local development doesn't look at the generated `index.d.ts` file for types like it will with the new `defined.d.ts` file

| Before                                                                                               | After                                                                                                  |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------ |
| ![2024-11-15_16-11](https://github.com/user-attachments/assets/c10f9963-ec8c-4b18-97df-ce5cf2ed940f) | ![2024-11-15_16-11_2](https://github.com/user-attachments/assets/0891ac05-2305-4fba-b2f0-9e520bd18b68) |


- Fixed `Check.d.ts` for local development so the TS server understands it's a default export not an exported object
    - This was originally added in https://github.com/CesiumGS/cesium/pull/11901 to help _external_ users of CesiumJS, primarily the ion team

| Before                                                                                               | After                                                                                                  |
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------ |
| ![2024-11-15_16-14](https://github.com/user-attachments/assets/ae89d7a3-6293-424c-9c9d-8038b0b08e13) | ![2024-11-15_16-19](https://github.com/user-attachments/assets/0eba7e61-b40d-4336-bd4e-3cecc06dca33) |

- Updated the build process to account for these files and changes. _The built `Cesium.d.ts` and `engine/index.d.ts` files should be identical to the previous ones_.







## Issue number and link

No issue, just a small DX issue I've grown too tired of seeing





## Testing plan

- In `main` run `npm run built-ts`
- Create a copy of `Source/Cesium.d.ts` and `packages/engine/index.d.ts` somewhere you can check later
- Switch to this branch and run `npm run build-ts`
- Do a diff of those files and ensure they are identical. This confirms our external use will be the same downstream
- To test locally find a usage of `defined` and `Check` locally
- Turn on the VSCode setting `Check JS` for type checking with the TS server (This may or may not be necessary, I'm not sure)
- Verify that in `main` the types aren't correct as shown above, verify that in this branch they are
- Turn off the `Check JS` setting if you're not ready to see a lot of other (false positive) errors scattered in our code



# Author checklist
